src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java

```diff
@@ -174,79 +174,105 @@ public static ClassDesc parseReferenceTypeDesc(String descriptor) {
+        // reject empty unqualified name or trailing separators
+        if (len == afterSeparator)
+            throw invalidClassName(name);
         return name;
     }
 
     /**
      * Validates the correctness of an internal class name.
-     * In particular checks for the presence of invalid characters in the name.
+     * In particular checks for the presence of invalid characters, empty
+     * name, consecutive, leading, or trailing {@code /}.
      *
      * @param name the class name
      * @return the class name passed if valid
      * @throws IllegalArgumentException if the class name is invalid
      * @throws NullPointerException if class name is {@code null}
      */
     public static String validateInternalClassName(String name) {
-        for (int i = 0; i < name.length(); i++) {
+        // state variable for detection of illegal states, such as:
+        // empty unqualified name, consecutive, leading, or trailing separators
+        int afterSeparator = 0;
+        int len = name.length();
+        for (int i = 0; i < len; i++) {
             char ch = name.charAt(i);
-            if (ch == ';' || ch == '[' || ch == '.'
-                    || ch == '/' && (i == 0 || i + 1 == name.length() || name.charAt(i - 1) == '/'))
+            // reject ';' or '[' or other form's separator
+            if (ch == ';' || ch == '[' || ch == '.')
                 throw invalidClassName(name);
+            if (ch == '/') {
+                // illegal state when received separator indicates consecutive
+                // or leading separators
+                if (i == afterSeparator)
+                    throw invalidClassName(name);
+                afterSeparator = i + 1;
+            }
         }
+        // reject empty unqualified name or trailing separators
+        if (len == afterSeparator)
+            throw invalidClassName(name);
         return name;
     }
 
     /**
      * Validates the correctness of a binary package name.
-     * In particular checks for the presence of invalid characters in the name.
-     * Empty package name is allowed.
+     * In particular checks for the presence of invalid characters, consecutive,
+     * leading, or trailing {@code .}.  Allows empty strings for the unnamed package.
      *
      * @param name the package name
      * @return the package name passed if valid
      * @throws IllegalArgumentException if the package name is invalid
      * @throws NullPointerException if the package name is {@code null}
      */
     public static String validateBinaryPackageName(String name) {
-        for (int i = 0; i < name.length(); i++) {
-            char ch = name.charAt(i);
-            if (ch == ';' || ch == '[' || ch == '/')
-                throw new IllegalArgumentException("Invalid package name: " + name);
-        }
-        return name;
+        // the unnamed package + null check
+        if (name.isEmpty())
+            return name;
+        return validateBinaryClassName(name);
```

Comment:
Perhaps have a utility method to specify if empty name is allowed.   `validateBinaryClassName` and `validateInternalClassName` only diff in the separator charactor.   They can be refactored to call one single name validation method something like this:

```
    private static String validateName(String name, boolean internalName, boolean allowEmptyName) {
        if (!allowEmptyName && name.isEmpty())
              throw invalidClassName(name);
              
        // state variable for detection of illegal states, such as:
        // empty unqualified name, consecutive, leading, or trailing separators
        int afterSeparator = 0;
        int len = name.length();
        for (int i = 0; i < len; i++) {
            char ch = name.charAt(i);
            // reject ';' or '[' or other form's separator
            if (ch == ';' || ch == '[' || (internalName && ch == '.') || (!internalName && ch == '/'))
                throw invalidClassName(name);
            if ((internalName && ch == '/') || (!internalName && ch == '.')) {
                // illegal state when received separator indicates consecutive
                // or leading separators
                if (i == afterSeparator)
                    throw invalidClassName(name);
                afterSeparator = i + 1;
            }
        }
        // reject empty unqualified name or trailing separators
        if (len == afterSeparator)
            throw invalidClassName(name);
        return name;
    }
```

Reply:
I assume I will still keep the 4 public methods to call this implementation method, as the boolean arguments will be otherwise confusing.  Is that what you intend?

Reply:
yes.  since you already made some refactoring, can improve further.